action/protocol/staking/protocol.go

```diff
@@ -109,3 +121,31 @@ func (p *Protocol) Register(r *protocol.Registry) error {
 func (p *Protocol) ForceRegister(r *protocol.Registry) error {
 	return r.ForceRegister(protocolID, p)
 }
+
+// CreateGenesisStates creates genesis states
+func (p *Protocol) CreateGenesisStates(ctx context.Context, sm protocol.StateManager) error {
+	if p.inMemCandidates.Size() > 0 {
+		return nil
+	}
+
+	// read all candidates from stateDB
+	stateDB := sm.GetDB()
+	candPkHash, candValue, err := stateDB.Filter(factory.CandidateNameSpace, func(k, v []byte) bool {
+		return bytes.HasPrefix(k, []byte{candidate})
+	})
+	if err != nil {
+		return err
+	}
+
+	// decode the candidate and put into candidate center
+	for i := range candPkHash {
+		c := &Candidate{}
+		if err := c.Deserialize(candValue[i]); err != nil {
+			return errors.Wrapf(err, "failed to deserialize candidate bytes %v", candValue[i])
+		}
+		if err := p.inMemCandidates.Put(c); err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
this is supposed to be in Start

Reply:
yes, but Start() does not have state manager